pilot/pkg/proxy/envoy/v2/discovery.go

```diff
@@ -499,32 +452,57 @@ func (s *DiscoveryServer) ConfigUpdate(full bool) {
 // Debouncing and update request happens in a separate thread, it uses locks
 // and we want to avoid complications, ConfigUpdate may already hold other locks.
 func (s *DiscoveryServer) handleUpdates() {
+	handleUpdates(s.updateChannel, DebounceAfter, DebounceMax, s.doPush)
+}
+
+// handleUpdates processes events from updateChannel
+// It ensures that at minimum minQuiet time has elapsed since the last event before processing it.
+// It also ensures that at most maxDelay is elapsed between receiving an event and processing it.
+func handleUpdates(updateChannel <-chan *updateReq, minQuiet time.Duration, maxDelay time.Duration, processUpdate func(bool)) {
+	var timeChan <-chan time.Time
+	var startDebounce time.Time
+	var lastConfigUpdateTime time.Time
+
+	configUpdateCounter := 0
+	pushCounter := 0
+
+	debouncedEvents := 0
+	fullPush := false
+
 	for {
 		select {
-		case r, _ := <-s.updateChannel:
-
-			if DebounceAfter == 0 {
-				go s.doPush()
-				continue
+		case r := <-updateChannel:
+			lastConfigUpdateTime = time.Now()
+			if debouncedEvents == 0 {
+				timeChan = time.After(minQuiet)
+				startDebounce = lastConfigUpdateTime
 			}
-			s.updateMutex.Lock()
-
+			debouncedEvents++
+			configUpdateCounter++
+			// fullPush is sticky if any debounced event requires a fullPush
 			if r.full {
-				s.fullPush = true
+				fullPush = true
 			}
-			s.configUpdateCounter++
 
-			s.lastConfigUpdateTime = time.Now()
-
-			if !s.debouncePushTimerSet {
-				s.debouncePushTimerSet = true
-				startDebounce := s.lastConfigUpdateTime
-				time.AfterFunc(DebounceAfter, func() {
-					s.debouncePush(startDebounce)
-				})
-			} // else: debounce in progress - it'll keep delaying the push
+		case now := <-timeChan:
+			timeChan = nil
+
+			eventDelay := now.Sub(startDebounce)
+			quietTime := now.Sub(lastConfigUpdateTime)
+			// it has been too long or quiet enough
+			if eventDelay >= maxDelay || quietTime >= minQuiet {
```

Comment:
do you even need this two way thing? Why not have a simpler logic that has a burstInterval - all events within this interval (since the first event) are squished and we do one config process at end of the burst interval.